### PR TITLE
Enable customization of columns in ItemsTable

### LINKED ISSUE #41908: [Browse] Make ItemsTable customizable

### PR DESCRIPTION
**Pull request by rafpaf**
Lets the `ItemsTable` component have customizable columns.

The logic for this is contained in `BaseItemsTable`. This is preliminary work for #41614

This doesn't introduce any new UI, but there is a unit test for the new logic.

The main thing to check is whether the table on the collections page stills work correctly. Note how the columns adjust as you resize the viewport horizontally. (In fact, the columns will sometimes adjust as you close and open the sidebar, because the width they are paying attention to is the width of the area to the right of the sidebar.)

Closes #41908

### CODE REVIEW

**Comment by rafpaf** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/41777?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#41614** <a href="https://app.graphite.dev/github/pr/metabase/metabase/41614?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#41777** <a href="https://app.graphite.dev/github/pr/metabase/metabase/41777?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#41756** <a href="https://app.graphite.dev/github/pr/metabase/metabase/41756?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#41746** <a href="https://app.graphite.dev/github/pr/metabase/metabase/41746?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rafpaf and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by rafpaf — COMMENTED** [review]
I've put this back into draft mode because I want to rethink the architecture. The approach could probably be simpler. Rather than use a `columns` prop, using a JSX expression would likely make the code more readable.

File: frontend/src/metabase/collections/components/BaseItemsTable.tsx [hunk 1]
```diff
@@ -13,72 +17,84 @@ import type {
...
       sort_direction: nextDirection,
     });
   };
 
   return (
-    <ColumnHeader>
+    <ColumnHeader
+      hideAtContainerBreakpoint={hideAtContainerBreakpoint}
+      containerName={containerName}
+    >
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BaseItemsTable.tsx:78]: On this branch, columns are hidden/shown depending on the width of the container. Rather than use the previously existing `col, th, td { &:nth-of-type ...` pattern, we add a `@container` query to certain cells in the table. This query sets `display: none` if the container shrinks below a given breakpoint.

File: frontend/src/metabase/collections/components/BulkActions.styled.tsx [hunk 2]
```diff
@@ -3,15 +3,12 @@ import styled from "@emotion/styled";
...
-import { NAV_SIDEBAR_WIDTH } from "metabase/nav/constants";
 import { space } from "metabase/styled-components/theme";
 
-export const BulkActionsToast = styled.div<{ isNavbarOpen: boolean }>`
+export const BulkActionsToast = styled.div`
   position: fixed;
   bottom: 0;
   left: 50%;
-  margin-left: ${props =>
-    props.isNavbarOpen ? `${parseInt(NAV_SIDEBAR_WIDTH) / 2}px` : "0"};
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BulkActions.styled.tsx:14]: This PR puts a [CSS container](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_containment/Container_queries) around the ItemsTable, so this adjustment is no longer needed (and actually makes the BulkActionsToast appear too far to the right).

File: frontend/src/metabase/components/ItemsTable/BaseItemsTable.styled.tsx [hunk 3]
```diff
@@ -6,81 +6,61 @@ import Link from "metabase/core/components/Link";
...
-        th:nth-of-type(${offsetEditedByIndex}),
-        col:nth-of-type(${offsetEditedByIndex}),
-        td:nth-of-type(${offsetEditedAtIndex}),
-        th:nth-of-type(${offsetEditedAtIndex}),
-        col:nth-of-type(${offsetEditedAtIndex}) {
-          display: none;
-        }
-      }
-    `;
-  }}
```
Comment (rafpaf) [inline frontend/src/metabase/components/ItemsTable/BaseItemsTable.styled.tsx:57]: Previously, columns were made responsive to viewport width changes through these hard-coded indexes. On this branch, the same effect is achieved with the help of `@container` queries

File: frontend/src/metabase/components/ItemsTable/BaseItemsTable.unit.spec.tsx [hunk 4]
```diff
@@ -75,7 +76,7 @@ describe("Collections BaseItemsTable", () => {
 
   it("displays item data", () => {
     setup();
-    const lastEditedAt = moment(timestamp).format("MMMM DD, YYYY");
+    const lastEditedAt = moment(timestamp).format("MMMM D, YYYY");
```
Comment (rafpaf) [inline frontend/src/metabase/components/ItemsTable/BaseItemsTable.unit.spec.tsx:79]: Since this branch uses `<DateTime unit="day" value={lastEditInfo.timestamp} />` to display the "last edited at" datetime, the formatting has now been corrected from "April 05, 2024" to "April 5, 2024", hence this change

File: frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx [hunk 5]
```diff
@@ -120,9 +119,11 @@ class DraggedItems extends Component {
       >
         <BaseItemsTable
           items={items}
-          renderItem={this.renderItem}
+          ItemComponent={props => this.renderItem(props)}
```
Comment (sloansparger) [inline frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx:122]: why is this needed?

**Review by sloansparger — APPROVED** [review]

**Comment by sloansparger** [comment]
despite the large change set this is a straightforward improvement, thanks for doing this!

File: frontend/src/metabase/components/EntityItem/EntityItem.styled.tsx [hunk 6]
```diff
@@ -42,8 +42,8 @@ export const EntityIconWrapper = styled(IconButtonWrapper)<{
 `;
 
 export const EntityItemWrapper = styled.div<{
-  variant: string;
-  disabled: boolean;
+  variant?: string;
+  disabled?: boolean;
```
[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/components/EntityItem/EntityItem.styled.tsx:46]: Just seems intuitive to me

File: frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx (hunk 5 above)
Comment (rafpaf) [reply frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx:122]: It seemed more conventional, and potentially less bug-prone, to pass a component rather than a function

File: frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx [hunk 7]
```diff
@@ -64,7 +64,7 @@ class DraggedItems extends Component {
     // necessary for decent drag performance
     return (
       nextProps.items.length !== this.props.items.length ||
-      nextProps.pinnedItems.length !== this.props.pinnedItems ||
+      nextProps.pinnedItems.length !== this.props.pinnedItems.length ||
```
Comment (rafpaf) [inline frontend/src/metabase/containers/dnd/ItemsDragLayer.jsx:67]: @sloansparger Just wanted to flag this new substantive change. Somehow there was a new dragging performance issue introduced in this branch, and this fixes it

File: frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx [hunk 8]
```diff
@@ -24,7 +24,7 @@ import type { State } from "metabase-types/store";
 
 import { EntityItemMenu } from "./ActionMenu.styled";
 
-interface OwnProps {
+export interface ActionMenuProps {
```
[the remark this answers is not in the document]
Comment (rafpaf) [reply frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:27]: I tried this but alas ActionMenuProps is needed in another file where those extra required props aren't expected

**Comment by rafpaf** [comment]
### Merge activity

* **May 3, 10:43 AM EDT**: @rafpaf merged this pull request with [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/41777).

**Comment by iethree** [comment]
yikes, this PR seems to have done something to mess with z-indexes on the collection page. investigating now.

Screen Shot 2024-05-07 at 10 39 45 AM
